src/pudl/transform/eia861.py

```diff
@@ -2327,76 +2146,290 @@ def utility_data(tfr_dfs):
+    This means that we have to hold off on the final normalization of the
+    balancing_authority_eia861 table until the rest of the transform process is over.
+    """
+    logger.info("Completing normalization of balancing_authority_eia861.")
+    ba_eia861_normed = clean_balancing_authority_eia861.loc[
+        :,
+        [
+            "report_date",
+            "balancing_authority_id_eia",
+            "balancing_authority_code_eia",
+            "balancing_authority_name_eia",
+        ],
+    ].drop_duplicates(subset=["report_date", "balancing_authority_id_eia"])
+
+    # Make sure that there aren't any more BA IDs we can recover from later years:
+    ba_ids_missing_codes = (
+        ba_eia861_normed.loc[
+            ba_eia861_normed.balancing_authority_code_eia.isnull(),
+            "balancing_authority_id_eia",
+        ]
+        .drop_duplicates()
+        .dropna()
+    )
+    fillable_ba_codes = ba_eia861_normed[
+        (ba_eia861_normed.balancing_authority_id_eia.isin(ba_ids_missing_codes))
+        & (ba_eia861_normed.balancing_authority_code_eia.notnull())
+    ]
+    if len(fillable_ba_codes) != 0:
+        raise ValueError(
+            f"Found {len(fillable_ba_codes)} unfilled but fillable BA Codes!"
         )
-        tfr_dfs[tfr_func] = _early_transform(raw_dfs[tfr_func])
-        tfr_dfs = tfr_funcs[tfr_func](tfr_dfs)
-
-    # This is more like harvesting stuff, and should probably be relocated:
-    tfr_dfs = balancing_authority_assn(tfr_dfs)
-    tfr_dfs = utility_assn(tfr_dfs)
-    tfr_dfs = normalize_balancing_authority(tfr_dfs)
-    # Do some final cleanup and assign types.
-    return {
-        name: convert_cols_dtypes(df, data_source="eia") for name, df in tfr_dfs.items()
-    }
+
+    return _post_process(ba_eia861_normed)
+
+
+def preclean_eia861_asset_factory(table_name: str) -> AssetsDefinition:
+    """An asset factory to apply uniform pre-processing to all raw EIA-861 assets."""
+    preclean_table_name = table_name.replace("raw_", "preclean_")
+
+    @asset(
+        ins={table_name: AssetIn()},
+        name=preclean_table_name,
+        group_name="eia861_preclean_assets",
+    )
+    def eia861_preclean_asset(**kwargs: dict[str, pd.DataFrame]) -> pd.DataFrame:
+        """Apply uniform pre-processing to all raw EIA-861 assets."""
+        return _pre_process(kwargs[table_name])
+
+    return eia861_preclean_asset
+
+
+def create_preclean_eia861_assets() -> list[AssetsDefinition]:
+    """Use the asset factory to generate all required preclean EIA-861 assets."""
+    table_names = [
+        asset_key.to_python_identifier()
+        for eia861_raw_asset in load_assets_from_modules([extract_eia861])
+        for asset_key in eia861_raw_asset.keys
+    ]
+    assets = []
+    for table_name in table_names:
+        assets.append(preclean_eia861_asset_factory(table_name))
+    return assets
+
+
+# eia861_preclean_assets = create_preclean_eia861_assets()
```

Comment:
I was a little bit surprised that just running this line was enough to cause all of the new assets to appear in the definition. What is it exactly that triggers that?

It also seemed like a bit of an anti-pattern to be calling this at the top level in the module, since whenever there was anything wrong with it, dagster would be unable to load the `pudl.etl` code location. But maybe this is normal and the way it's done?

Reply:
From what I can tell this how things are done. `load_assets_from_modules()` just looks for `AssetDefinitions` floating around in the module.